src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,287 @@
+#define Z180_STAT0_TIE          0x01
+
+/* 05 ASCI status register 1 (all bits read-only except RIE, CTS1E and TIE) */
+#define Z180_STAT1_RDRF         0x80
+#define Z180_STAT1_OVRN         0x40
+#define Z180_STAT1_PE           0x20
+#define Z180_STAT1_FE           0x10
+#define Z180_STAT1_RIE          0x08
+#define Z180_STAT1_CTS1E        0x04
+#define Z180_STAT1_TDRE         0x02
+#define Z180_STAT1_TIE          0x01
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(Z180ASCI_CHANNEL,   z180asci_channel,  "z180asci_channel",    "Z180 ASCI Channel")
+
+
+//-------------------------------------------------
+//  z180asci_channel - constructor
+//-------------------------------------------------
+
+z180asci_channel::z180asci_channel(
+		const machine_config &mconfig,
+		device_type type,
+		const char *tag,
+		device_t *owner,
+		uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock),
+	device_buffered_serial_interface(mconfig, *this),
+	m_tx_handler(*this)
+{
+	m_asci_cntla = 0;
+	m_asci_cntlb = 7; // SS2,SS1 and SS0 are 1 after reset
+	m_asci_stat = 2; // set TDRE=1 on reset
+	m_asci_tdr = 0;
+	m_asci_rdr = 0;
+	m_id = 0;
+	m_irq = 0;
+}
+
+z180asci_channel::z180asci_channel(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: z180asci_channel(mconfig, Z180ASCI_CHANNEL, tag, owner, clock)
+{
+}
+
+//-------------------------------------------------
+//  device_resolve_objects - resolve objects that
+//  may be needed for other devices to set
+//  initial conditions at start time
+//-------------------------------------------------
+
+void z180asci_channel::device_resolve_objects()
+{
+	// resolve callbacks
+	m_tx_handler.resolve_safe();
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void z180asci_channel::device_start()
+{
+	save_item(NAME(m_asci_cntla));
+	save_item(NAME(m_asci_cntlb));
+	save_item(NAME(m_asci_stat));
+	save_item(NAME(m_asci_tdr));
+	save_item(NAME(m_asci_rdr));
+
+	transmit_register_reset();
+	receive_register_reset();
+
+	cpu_device *cpu = reinterpret_cast<cpu_device *>(owner());
+	if (m_id==0) {
+		cpu->state_add(Z180_CNTLA0,     "CNTLA0",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB0,     "CNTLB0",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT0,      "STAT0",     m_asci_stat);
+		cpu->state_add(Z180_TDR0,       "TDR0",      m_asci_tdr);
+		cpu->state_add(Z180_RDR0,       "RDR0",      m_asci_rdr);
+	} else {
+		cpu->state_add(Z180_CNTLA1,     "CNTLA1",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB1,     "CNTLB1",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT1,      "STAT1",     m_asci_stat);
+		cpu->state_add(Z180_TDR1,       "TDR1",      m_asci_tdr);
+		cpu->state_add(Z180_RDR1,       "RDR1",      m_asci_rdr);
+	}
```

Comment:
The broken encapsulation here is kind of the core's fault.  It would be useful for things that weren't the main CPU device to be able to add registers to the debugger's register pane.  Particularly in highly-integrated SoCs.  But that would be a fair amount of work across all of the debuggers, I believe.